diff --git a/lang/c/CMakeLists.txt b/lang/c/CMakeLists.txt
index 6c8d7aaf..f4e93e3d 100644
--- a/lang/c/CMakeLists.txt
+++ b/lang/c/CMakeLists.txt
@@ -149,11 +149,10 @@ else (ZLIB_FOUND)
     message("Disabled deflate codec. zlib not found.")
 endif (ZLIB_FOUND)
 
-find_package(Snappy)
+find_package(Snappy CONFIG)
 if (SNAPPY_FOUND AND ZLIB_FOUND)  # Snappy borrows crc32 from zlib
     set(SNAPPY_PKG libsnappy)
     add_definitions(-DSNAPPY_CODEC)
-    include_directories(${SNAPPY_INCLUDE_DIRS})
     message("Enabled snappy codec")
 else (SNAPPY_FOUND AND ZLIB_FOUND)
     set(SNAPPY_PKG "")
@@ -175,7 +174,7 @@ else (LZMA_FOUND)
     message("Disabled lzma codec. liblzma not found.")
 endif (LZMA_FOUND)
 
-set(CODEC_LIBRARIES ${ZLIB_LIBRARIES} ${LZMA_LIBRARIES} ${SNAPPY_LIBRARIES})
+set(CODEC_LIBRARIES ${ZLIB_LIBRARIES} ${LZMA_LIBRARIES} Snappy::snappy)
 set(CODEC_PKG "@ZLIB_PKG@ @LZMA_PKG@ @SNAPPY_PKG@")
 
 # Jansson JSON library
diff --git a/lang/c/FindSnappy.cmake b/lang/c/FindSnappy.cmake
index 5f158fdc..811cbe01 100644
--- a/lang/c/FindSnappy.cmake
+++ b/lang/c/FindSnappy.cmake
@@ -36,7 +36,7 @@
 #  SNAPPY_INCLUDE_DIR        The location of Snappy headers
 
 find_path(SNAPPY_INCLUDE_DIR
-    NAMES snappy.h
+    NAMES snappy-c.h
     HINTS ${SNAPPY_ROOT_DIR}/include)
 
 find_library(SNAPPY_LIBRARIES
